### ISSUE TITLE
Controlling tag output

### ISSUE THREAD

**Issue by izwick-schachter** [lsegal/yard#1150]
I'm attempting to create a custom YARD tag, and control the HTML which is outputted by that tag. I've spent multiple hours combing documentation looking for how I can accomplish this (the most useful being the current docs for [Tags::Library](http://www.rubydoc.info/gems/yard/YARD/Tags/Library) and [in an older version](http://www.rubydoc.info/gems/yard/0.7.1/YARD/Tags/Library)), but I've still been unsuccessful.

I've managed to define a new tag which does show up in my docs, but have not been able to control the output format or anything beyond that. So, how can I specify the HTML to be outputted by YARD for a tag I've defined?

Once I've figured it out, I'll open a pull request to update the documentation to make more sense.

## Steps to reproduce

I've been working with rake (version 12.3.0), so here is my Rakefile:

```ruby
require 'yard'

class MyTagFactory < YARD::Tags::DefaultFactory
  def parse_tag(tag, content)
    YARD::Tags::Tag.new(tag, "My Tag Content: #{content}")
  end
end

YARD::Tags::Library.default_factory = MyTagFactory.new
YARD::Tags::Library.visible_tags += [:mytag]
YARD::Tags::Library.define_tag("My Tag", :mytag)

YARD::Rake::YardocTask.new(:yard)
```

And, the contents of `lib/yard.rb`:

```ruby
# @mytag Testing, this is what I want outputted
def foo
end
```

And, (in case it's relevant) here's my Gemfile:

```ruby
source 'https://rubygems.org'

gem 'yard'
gem 'rake'
```

Then, to generate the documentation, I ran `bundle exec rake yard`, which output:

```
Files:           1
Modules:         0 (    0 undocumented)
Classes:         0 (    0 undocumented)
Constants:       0 (    0 undocumented)
Attributes:      0 (    0 undocumented)
Methods:         1 (    0 undocumented)
 100.00% documented
```

## Actual Output

Here's the output I got (viewing `doc/top-level-namespace.html#foo-instance_method` in my browser):

# My Tag:

- My Tag Content: Testing, this is what I want outputted

## Expected Output

I'd like it to simply output (in browser) what was input, with no formatting: "My Tag Content: Testing, this is what I want outputted"

## Environment details:

* OS: MacOS Sierra 10.12.6
* Ruby version (`ruby -v`): ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-darwin16]
* YARD version (`yard -v`): yard 0.9.12
* Rake version (`rake --version`): rake, version 12.3.0
* Browser: Google Chrome 63.0.3239.84 (Official Build) (64-bit)

I have read the [Contributing Guide][contrib].

[contrib]: https://github.com/lsegal/yard/blob/master/CONTRIBUTING.md

**Comment by izwick-schachter** [comment]
More specifically, I have a custom tag (`@spec`) which I would like to generate something that looks like `@note` containing the parent class name and `{include:file:<filename>}` where `<filename>` is chosen based on the method name.

**Comment by lsegal** [comment]
You can use `--tag mytag:"My Tag"` in your `.yardopts` file to achieve this, mostly. If you need more customization than the `--*tag*` options provide, you can customize using [templating](https://github.com/lsegal/yard/blob/master/docs/Templates.md).

**Comment by izwick-schachter** [comment]
I looked at templating, and it seemed to me that templating was for doing things based on the code structure. In my case, I want to simply add a programmatically generated note to all methods that have a certain tag.

**Comment by lsegal** [comment]
I'm not sure I quite understand, but if you are necessarily looking for a separate tag to be semantically added to your parse results, you can use the [YARD::SourceParser.after_parse_list](http://www.rubydoc.info/gems/yard/YARD/Parser/SourceParser#after_parse_list-class_method) hook to programmatically do things with the YARD database like add new tags. Example:

```
YARD::SourceParser.after_parse_list do
  YARD::Registry.all(:method).each do |meth|
    next unless meth.has_tag?(:spec)
    meth.add_tag YARD::Tags::Tag.new(:note, 'note contents')
  end
end
```

**Comment by izwick-schachter** [comment]
For anyone who stumbles upon this in the future, here is what I came up with. It's exactly what you suggested, with a couple tweaks:

```ruby
YARD::Tags::Library.define_tag('MyTag', :mytag)
YARD::Parser::SourceParser.after_parse_list do
  YARD::Registry.all(:method).each do |meth|
    next unless meth.has_tag?(:mytag)
    # A useful thing
    tag = meth.tag(:mytag)
    meth.add_tag YARD::Tags::Tag.new(:note, "This tag is applied to #{meth}")
  end
end
```

Thank you for your help @lsegal, and thank you for YARD!

**Comment by izwick-schachter** [comment]
Oops, I forgot something. Would it be appropriate to document this in the docs for `YARD::Parser::SourceParser.after_parse_list`? If so, I'll open a pull request.

**Comment by lsegal** [comment]
@izwick-schachter correct me if I'm wrong but it seems like the issue here was more about visibility of the `after_parse_list` hook method (you didn't know that this functionality existed), and not necessarily how to use the method once found. If so, the documentation solution here would be to add visibility to the hook in one of the high level guide style docs rather than just adding an example to `after_parse_list`. 

What do you think?

**Comment by izwick-schachter** [comment]
I agree that `after_parse_list` should be more visible, but even if I'd found the documentation for it I still wouldn't have made the connection to my use case. So, I think that it should get added to one of the high level guides and also get a section added describing use cases similar to mine.

**Comment by lsegal** [comment]
> [!NOTE]
> This is a batch message, apologies if this feels impersonal.

I am closing out a few stale issues and PRs in the YARD repo and this was flagged for automatic closure because the issue may have been fixed or is likely not planned due to the issue/PR age. If this is still an issue or you believe the issue/PR is still worth a discussion, please feel free to comment or reopen a new issue. Thanks and sorry for not getting around to this the first time!
